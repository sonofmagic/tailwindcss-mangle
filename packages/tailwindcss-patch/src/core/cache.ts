import fs from 'node:fs'
import path from 'node:path'
import { pkgName } from '@/constants'
import type { CacheOptions, InternalCacheOptions } from '@/types'
import { log } from '@/logger'

export function getCacheOptions(options?: CacheOptions | boolean) {
  let cache: InternalCacheOptions
  switch (typeof options) {
    case 'undefined': {
      cache = {
        enable: false,
      }
      break
    }
    case 'boolean': {
      cache = {
        enable: options,
      }
      break
    }
    case 'object': {
      cache = { ...options, enable: true }
      break
    }
  }
  return cache
}

export class CacheManager {
  public options: Required<CacheOptions> & { filename: string }

  constructor(options: CacheOptions = {}) {
    this.options = this.getOptions(options)
  }

  mkdir(cacheDirectory: string) {
    const exists = fs.existsSync(cacheDirectory)
    if (!exists) {
      fs.mkdirSync(cacheDirectory, {
        recursive: true,
      })
    }
    return cacheDirectory
  }

  getOptions(options: CacheOptions = {}): Required<CacheOptions> & { filename: string } {
    const cwd = options.cwd ?? process.cwd()
    const dir = options.dir ?? path.resolve(cwd, 'node_modules/.cache', pkgName)
    const file = options.file ?? 'index.json'
    const filename = path.resolve(dir, file)
    return {
      cwd,
      dir,
      file,
      filename,
      strategy: 'merge',
    }
  }

  write(data: Set<string>) {
    try {
      const { dir, filename } = this.options
      this.mkdir(dir)
      fs.writeFileSync(filename, JSON.stringify([...data], undefined, 2), 'utf8')
      return filename
    }
    catch {
      log('write cache file fail!')
    }
  }

  read() {
    const { filename } = this.options
    try {
      if (fs.existsSync(filename)) {
        const data = fs.readFileSync(filename, 'utf8')
        return new Set<string>(JSON.parse(data))
      }
    }
    catch {
      log(`parse cache content fail! path:${filename}`)
      try {
        fs.unlinkSync(filename)
      }
      catch {
        log(`delete cache file fail! path:${filename}`)
      }
    }
  }
}
