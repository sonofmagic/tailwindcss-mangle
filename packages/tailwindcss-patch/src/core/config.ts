export * from '@tailwindcss-mangle/config'
