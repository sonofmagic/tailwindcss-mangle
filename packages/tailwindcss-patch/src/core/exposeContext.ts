import path from 'node:path'
import fs from 'node:fs'
import type { TailwindcssClassCache, TailwindcssRuntimeContext } from '@/types'
import { requireResolve } from '@/utils'

export function getTailwindcssEntry(basedir?: string) {
  return requireResolve('tailwindcss', {
    basedir,
  })
}

export function getContexts(basedir?: string): TailwindcssRuntimeContext[] {
  const twPath = getTailwindcssEntry(basedir)

  const distPath = path.dirname(twPath)

  let injectFilePath = path.join(distPath, 'plugin.js')
  if (!fs.existsSync(injectFilePath)) {
    injectFilePath = path.join(distPath, 'index.js')
  }

  const mo = require(injectFilePath)
  if (mo.contextRef) {
    return mo.contextRef.value
  }
  return []
}

export function getClassCaches(basedir?: string): TailwindcssClassCache[] {
  const contexts = getContexts(basedir)
  return contexts.map(x => x.classCache)
}

export function getClassCacheSet(basedir?: string, options?: { removeUniversalSelector?: boolean }): Set<string> {
  const classCaches = getClassCaches(basedir)
  const classSet = new Set<string>()
  for (const classCacheMap of classCaches) {
    const keys = classCacheMap.keys()
    for (const key of keys) {
      const v = key.toString()
      if (options?.removeUniversalSelector && v === '*') {
        continue
      }
      classSet.add(v)
    }
  }
  return classSet
}
