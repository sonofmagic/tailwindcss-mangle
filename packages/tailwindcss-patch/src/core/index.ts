export * from './tw-patcher'
export * from './exposeContext'
export * from './inspector'
export * from './runtime-patcher'
export * from './cache'
export * from './config'
