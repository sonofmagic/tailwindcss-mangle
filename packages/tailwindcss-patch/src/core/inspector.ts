import * as t from '@babel/types'
import { parse, traverse, generate } from '@/babel'
// crash code

export function inspectProcessTailwindFeaturesReturnContext(content: string) {
  const ast = parse(content)
  let hasPatched = false
  traverse(ast, {
    FunctionDeclaration(p) {
      const n = p.node
      if (n.id?.name === 'processTailwindFeatures' && n.body.body.length === 1 && t.isReturnStatement(n.body.body[0])) {
        const rts = n.body.body[0]
        if (t.isFunctionExpression(rts.argument)) {
          const body = rts.argument.body.body
          const lastStatement = body.at(-1)
          hasPatched = t.isReturnStatement(lastStatement) && t.isIdentifier(lastStatement.argument) && lastStatement.argument.name === 'context'
          if (!hasPatched) {
            // return context;
            const rts = t.returnStatement(t.identifier('context'))
            body.push(rts)
          }
        }
      }
    }
  })

  return {
    code: hasPatched ? content : generate(ast).code,
    hasPatched
  }
}

export function inspectPostcssPlugin(content: string) {
  const ast = parse(content)
  const exportKey = 'contextRef'
  const variableName = 'contextRef'
  const valueKey = 'value'
  let hasPatched = false
  traverse(ast, {
    Program(p) {
      const n = p.node
      // find module.exports = function tailwindcss(configOrPath)
      const idx = n.body.findIndex((x) => {
        return (
          t.isExpressionStatement(x) &&
          t.isAssignmentExpression(x.expression) &&
          t.isMemberExpression(x.expression.left) &&
          t.isFunctionExpression(x.expression.right) &&
          x.expression.right.id?.name === 'tailwindcss'
        )
      })

      if (idx > -1) {
        const prevStatement = n.body[idx - 1]
        const lastStatement = n.body.at(-1)
        const hasPatchedCondition0 =
          prevStatement &&
          t.isVariableDeclaration(prevStatement) &&
          prevStatement.declarations.length === 1 &&
          t.isIdentifier(prevStatement.declarations[0].id) &&
          prevStatement.declarations[0].id.name === variableName
        const hasPatchedCondition1 =
          t.isExpressionStatement(lastStatement) &&
          t.isAssignmentExpression(lastStatement.expression) &&
          t.isIdentifier(lastStatement.expression.right) &&
          lastStatement.expression.right.name === variableName

        hasPatched = hasPatchedCondition0 || hasPatchedCondition1
        if (!hasPatched) {
          // const contextRef = {
          //   value: []
          // };
          const statement = t.variableDeclaration('const', [
            t.variableDeclarator(t.identifier(variableName), t.objectExpression([t.objectProperty(t.identifier(valueKey), t.arrayExpression())]))
          ])
          n.body.splice(idx, 0, statement)
          // module.exports.contextRef = contextRef;
          n.body.push(
            t.expressionStatement(
              t.assignmentExpression(
                '=',
                t.memberExpression(t.memberExpression(t.identifier('module'), t.identifier('exports')), t.identifier(exportKey)),
                t.identifier(variableName)
              )
            )
          )
        }
      }
    },
    FunctionExpression(p) {
      if (hasPatched) {
        return
      }
      const n = p.node
      if (n.id?.name === 'tailwindcss' && n.body.body.length === 1 && t.isReturnStatement(n.body.body[0])) {
        const returnStatement = n.body.body[0]
        if (t.isObjectExpression(returnStatement.argument) && returnStatement.argument.properties.length === 2) {
          const properties = returnStatement.argument.properties
          if (t.isObjectProperty(properties[0]) && t.isObjectProperty(properties[1])) {
            const keyMatched = t.isIdentifier(properties[0].key) && properties[0].key.name === 'postcssPlugin'
            const pluginsMatched = t.isIdentifier(properties[1].key) && properties[1].key.name === 'plugins'
            if (
              pluginsMatched &&
              keyMatched &&
              t.isCallExpression(properties[1].value) &&
              t.isMemberExpression(properties[1].value.callee) &&
              t.isArrayExpression(properties[1].value.callee.object)
            ) {
              const pluginsCode = properties[1].value.callee.object.elements
              if (pluginsCode[1] && t.isFunctionExpression(pluginsCode[1])) {
                const targetBlockStatement = pluginsCode[1].body

                const lastStatement = targetBlockStatement.body.at(-1)
                if (t.isExpressionStatement(lastStatement)) {
                  // contextRef.value.push((0, _processTailwindFeatures.default)(context)(root, result));
                  const newExpressionStatement = t.expressionStatement(
                    t.callExpression(
                      t.memberExpression(
                        t.memberExpression(t.identifier(variableName), t.identifier('value')),

                        t.identifier('push')
                      ),

                      [lastStatement.expression]
                    )
                  )
                  targetBlockStatement.body[targetBlockStatement.body.length - 1] = newExpressionStatement
                }

                const ifIdx = targetBlockStatement.body.findIndex((x) => t.isIfStatement(x))
                if (ifIdx > -1) {
                  const ifRoot = <t.IfStatement>targetBlockStatement.body[ifIdx]
                  if (t.isBlockStatement(ifRoot.consequent) && ifRoot.consequent.body[1] && t.isForOfStatement(ifRoot.consequent.body[1])) {
                    const forOf: t.ForOfStatement = ifRoot.consequent.body[1]
                    if (t.isBlockStatement(forOf.body) && forOf.body.body.length === 1 && t.isIfStatement(forOf.body.body[0])) {
                      const if2: t.IfStatement = forOf.body.body[0]
                      if (t.isBlockStatement(if2.consequent) && if2.consequent.body.length === 1 && t.isExpressionStatement(if2.consequent.body[0])) {
                        const target = if2.consequent.body[0]
                        // contextRef.value.push((0, _processTailwindFeatures.default)(context)(root1, result));
                        const newExpressionStatement = t.expressionStatement(
                          t.callExpression(t.memberExpression(t.memberExpression(t.identifier(variableName), t.identifier('value')), t.identifier('push')), [target.expression])
                        )
                        if2.consequent.body[0] = newExpressionStatement
                      }
                    }
                  }
                }
                // clear contextRef.value
                targetBlockStatement.body.unshift(
                  // contentRef.value = []
                  // t.expressionStatement(t.assignmentExpression('=', t.memberExpression(t.identifier(variableName), t.identifier(valueKey)), t.arrayExpression()))

                  // contentRef.value.length = 0
                  t.expressionStatement(
                    t.assignmentExpression(
                      '=',
                      t.memberExpression(t.memberExpression(t.identifier(variableName), t.identifier(valueKey)), t.identifier('length')),
                      t.numericLiteral(0)
                    )
                  )
                )
              }
            }
          }
        }
      }
      // start = true
    }
    // BlockStatement(p) {
    //   const n = p.node
    //   if (start && p.parent.type === 'FunctionExpression' && !p.parent.id) {
    //     n.body.unshift(t.expressionStatement(t.assignmentExpression('=', t.memberExpression(t.identifier(variableName), t.identifier(valueKey)), t.arrayExpression())))
    //   }
    // }
  })
  return {
    code: hasPatched ? content : generate(ast).code,
    hasPatched
  }
}
