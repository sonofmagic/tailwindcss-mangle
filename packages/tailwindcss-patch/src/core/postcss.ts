import postcss from 'postcss'
import path from 'node:path'

export async function processTailwindcss(options: { cwd?: string; config?: string }) {
  options.cwd = options.cwd ?? process.cwd()
  const config = options.config && path.isAbsolute(options.config) ? options.config : path.resolve(options.cwd, 'tailwind.config.js')

  const res = await postcss([
    require('tailwindcss')({
      config
    })
  ]).process('@tailwind base;@tailwind components;@tailwind utilities;', {
    from: undefined
  })
  return res
}
