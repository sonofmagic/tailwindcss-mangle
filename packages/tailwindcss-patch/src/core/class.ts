import { getClassCacheSet, getContexts, getTailwindcssEntry } from './exposeContext'
import type { CacheOptions, InternalCacheOptions, InternalPatchOptions, TailwindcssPatcherOptions, CacheStrategy } from '../types'
import { writeCache, readCache } from './cache'
import { createPatch, getPatchOptions } from './patcher'

export function getCacheOptions(options?: CacheOptions | boolean) {
  let cache: InternalCacheOptions
  switch (typeof options) {
    case 'undefined': {
      cache = {
        enable: false
      }
      break
    }
    case 'boolean': {
      cache = {
        enable: options
      }
      break
    }
    case 'object': {
      cache = { ...options, enable: true }
      break
    }
  }
  return cache
}
export class TailwindcssPatcher {
  public rawOptions: TailwindcssPatcherOptions
  public cacheOptions: InternalCacheOptions
  public patchOptions: InternalPatchOptions
  public patch: () => void
  constructor(options: TailwindcssPatcherOptions = {}) {
    this.rawOptions = options
    this.cacheOptions = getCacheOptions(options.cache)
    this.patchOptions = getPatchOptions(options.patch)
    this.patch = createPatch(this.patchOptions)
  }

  getPkgEntry(basedir?: string) {
    return getTailwindcssEntry(basedir)
  }

  setCache(set: Set<string>) {
    if (this.cacheOptions.enable) {
      return writeCache(set, this.cacheOptions)
    }
  }

  getCache() {
    // if(this.cache.enable){
    return readCache(this.cacheOptions)
    // }
  }

  /**
   * @description 在多个 tailwindcss 上下文时，这个方法将被执行多次，所以策略上应该使用 append
   * 详见 taro weapp-tailwindcss 独立分包
   * @param basedir
   * @returns
   */
  getClassSet(
    options: {
      basedir?: string
      cacheStrategy?: CacheStrategy
    } = {
      cacheStrategy: this.cacheOptions.strategy ?? 'merge'
    }
  ) {
    const { basedir, cacheStrategy } = options
    const set = getClassCacheSet(basedir)
    if (cacheStrategy === 'overwrite') {
      set.size > 0 && this.setCache(set)
    } else if (cacheStrategy === 'merge') {
      const cacheSet = this.getCache()
      if (cacheSet) {
        for (const x of cacheSet) {
          set.add(x)
        }
      }
      this.setCache(set)
    }

    return set
  }

  getContexts(basedir?: string) {
    return getContexts(basedir)
  }
}
