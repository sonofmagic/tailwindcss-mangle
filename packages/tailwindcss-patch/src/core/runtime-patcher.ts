import path from 'node:path'
import fs from 'node:fs'
import { gte } from 'semver'
import { inspectPostcssPlugin, inspectProcessTailwindFeaturesReturnContext } from './inspector'
import type { PatchOptions, InternalPatchOptions } from '@/types'
import type { PackageJson } from 'pkg-types'
import { defu } from 'defu'
import { defaultOptions } from '@/defaults'
import { ensureFileContent, requireResolve } from '@/utils'

export function getInstalledPkgJsonPath(options: PatchOptions = {}) {
  try {
    // const cwd = process.cwd()
    const tmpJsonPath = requireResolve(`tailwindcss/package.json`, {
      paths: options.paths
    })

    return tmpJsonPath
    // https://github.com/sonofmagic/weapp-tailwindcss-webpack-plugin
    // only tailwindcss version > 3.0.0
  } catch (error) {
    if ((<Error & { code: string }>error).code === 'MODULE_NOT_FOUND') {
      console.warn("Can't find npm pkg: `tailwindcss`, Please ensure it has been installed!")
    }
  }
}

export function getPatchOptions(options: PatchOptions = {}) {
  return defu(
    options,
    {
      basedir: process.cwd()
    },
    defaultOptions
  ) as InternalPatchOptions
}

export function createPatch(opt: InternalPatchOptions) {
  return () => {
    try {
      const pkgJsonPath = getInstalledPkgJsonPath(opt)
      return internalPatch(pkgJsonPath, opt)
    } catch (error) {
      console.warn(`patch tailwindcss failed:` + (<Error>error).message)
    }
  }
}

export function monkeyPatchForExposingContext(twDir: string, opt: InternalPatchOptions) {
  const processTailwindFeaturesFilePath = path.resolve(twDir, 'lib/processTailwindFeatures.js')

  const processTailwindFeaturesContent = ensureFileContent(processTailwindFeaturesFilePath)
  const result: { processTailwindFeatures?: string; plugin?: string } & Record<string, any> = {}
  if (processTailwindFeaturesContent) {
    const { code, hasPatched } = inspectProcessTailwindFeaturesReturnContext(processTailwindFeaturesContent)
    if (!hasPatched && opt.overwrite) {
      fs.writeFileSync(processTailwindFeaturesFilePath, code, {
        encoding: 'utf8'
      })
      console.log('patch tailwindcss processTailwindFeatures for return content successfully!')
    }
    result.processTailwindFeatures = code
  }

  const pluginFilePath = path.resolve(twDir, 'lib/plugin.js')
  const indexFilePath = path.resolve(twDir, 'lib/index.js')
  const pluginContent = ensureFileContent([pluginFilePath, indexFilePath])
  if (pluginContent) {
    const { code, hasPatched } = inspectPostcssPlugin(pluginContent)
    if (!hasPatched && opt.overwrite) {
      fs.writeFileSync(pluginFilePath, code, {
        encoding: 'utf8'
      })
      console.log('patch tailwindcss for expose runtime content successfully!')
    }
    result.plugin = code
  }

  opt.custom && typeof opt.custom === 'function' && opt.custom(twDir, result)
  return result
}

export function internalPatch(pkgJsonPath: string | undefined, options: InternalPatchOptions): any | undefined {
  if (pkgJsonPath) {
    const pkgJson = require(pkgJsonPath) as PackageJson
    const twDir = path.dirname(pkgJsonPath)
    if (gte(pkgJson.version!, '3.0.0')) {
      options.version = pkgJson.version
      const result = monkeyPatchForExposingContext(twDir, options)
      return result
    }
    // no sth
  }
}
