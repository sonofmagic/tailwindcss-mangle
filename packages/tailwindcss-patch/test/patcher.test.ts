import { getInstalledPkgJsonPath } from '../src/core/patcher'
import path from 'node:path'

describe('patcher', () => {
  it('getInstalledPkgJsonPath common options', () => {
    const pkgJsonPath = getInstalledPkgJsonPath()
    expect(pkgJsonPath).toBeTruthy()
    pkgJsonPath && expect(path.relative(process.cwd(), pkgJsonPath).replaceAll('\\', '/')).toContain('node_modules/tailwindcss/package.json')
  })

  it('getInstalledPkgJsonPath baseDir options', () => {
    const pkgJsonPath = getInstalledPkgJsonPath({
      basedir: path.resolve(__dirname, '../../../')
    })
    expect(pkgJsonPath).toBeTruthy()
  })
})
