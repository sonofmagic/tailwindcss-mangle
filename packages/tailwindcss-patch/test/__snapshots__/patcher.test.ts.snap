// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`patcher > getInstalledPkgJsonPath common options 1`] = `"node_modules/tailwindcss/package.json"`;
